### PR TITLE
Add support for gitlab.com nested groups in sourceUrl discovery

### PR DESCRIPTION
**Pull request by otbe**
## Changes

This PR adds support for nested groups of gitlab.com in the discovery/parsing of the sourceUrl of a dependency update.

## Context

See https://github.com/renovatebot/renovate/discussions/26721

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/metadata.ts [hunk 1]
```diff
@@ -233,3 +238,13 @@ export function shouldDeleteHomepage(
...
 }
+
+function tryParseGitlabUrl(url: string): string | null {
+  const result = parseGitlabUrl(url);
+
+  if (result === null || result.type !== 'gitlab') {
+    return null;
+  }
+
+  return `https://gitlab.com/${result.owner}/${result.name}`;
```
Comment (viceice) [inline lib/modules/datasource/metadata.ts:249]: this seems also wrong. what if parsed tape is gitlab but different host (selfhosted)?

**Review by viceice — CHANGES_REQUESTED** [review]
I think this needs a refactor before. the new code makes the whole massaging more worse .

we need to think about it more before.

File: lib/modules/datasource/metadata.spec.ts [hunk 2]
```diff
@@ -181,6 +181,54 @@ describe('modules/datasource/metadata', () => {
...
+      sourceUrl: 'https://gitlab.com/my/group/repo.git',
+      releases: [
+        { version: '1.0.0', releaseTimestamp: '2020-02-14T13:12:00.000Z' },
+      ],
+    };
+    const datasource = NpmDatasource.id;
+    const packageName = 'repo';
+
+    addMetaData(dep, datasource, packageName);
+    expect(dep).toMatchSnapshot({
```
[the remark this answers is not in the document]
Comment (otbe) [reply lib/modules/datasource/metadata.spec.ts]: Switched to `toMatchObject`. I just copied one of the other tests :D

File: lib/modules/datasource/metadata.spec.ts [hunk 3]
```diff
@@ -181,6 +181,54 @@ describe('modules/datasource/metadata', () => {
...
+      sourceUrl: 'git+ssh://gitlab.com/my/group/repo.git',
+      releases: [
+        { version: '1.0.0', releaseTimestamp: '2020-02-14T13:12:00.000Z' },
+      ],
+    };
+    const datasource = NpmDatasource.id;
+    const packageName = 'repo';
+
+    addMetaData(dep, datasource, packageName);
+    expect(dep).toMatchSnapshot({
```
[the remark this answers is not in the document]
Comment (otbe) [reply lib/modules/datasource/metadata.spec.ts]: Switched to `toMatchObject`. I just copied one of the other tests :D

File: lib/modules/datasource/metadata.spec.ts [hunk 4]
```diff
@@ -181,6 +181,54 @@ describe('modules/datasource/metadata', () => {
...
+      sourceUrl: 'git+ssh://foo:bar@gitlab.com/my/group/repo.git',
+      releases: [
+        { version: '1.0.0', releaseTimestamp: '2020-02-14T13:12:00.000Z' },
+      ],
+    };
+    const datasource = NpmDatasource.id;
+    const packageName = 'repo';
+
+    addMetaData(dep, datasource, packageName);
+    expect(dep).toMatchSnapshot({
```
[the remark this answers is not in the document]
Comment (otbe) [reply lib/modules/datasource/metadata.spec.ts]: Switched to `toMatchObject`. I just copied one of the other tests :D

File: lib/modules/datasource/metadata.ts (hunk 1 above)
Comment (otbe) [reply lib/modules/datasource/metadata.ts:249]: I think right nowthe library only works for certain official domains. This function only tries to mitigate the issue discussed in https://github.com/renovatebot/renovate/discussions/26721 without introducing any additional/changed behaviour :)
Thats also why `parse()` is still used as the front runner.

File: lib/modules/datasource/metadata.ts [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import is from '@sindresorhus/is';
 import parse from 'github-url-from-git';
 import { DateTime } from 'luxon';
+import parseGitlabUrl from 'parse-git-url';
```
[the remark this answers is not in the document]
Comment (otbe) [reply lib/modules/datasource/metadata.ts:4]: Point is I dont want to introduce a generic `parseGitUrl` here. All of this is so fragile and error prone :D

File: lib/modules/datasource/metadata.ts (hunk 1 above)
Comment (rarkins) [reply lib/modules/datasource/metadata.ts:249]: @otbe how wide is this intended to be? e.g. solving for gitlab.com only?

**Pull request closed without merging**